### PR TITLE
make sure that job.commit is always specified

### PR DESCRIPTION
**Pull request by zehua**
After the change in https://github.com/zendesk/samson/pull/1082, Samson fails to deploy from a commit hash, with the following error:

```
AIRBRAKE: NoMethodError - undefined method `shellescape' for nil:NilClass - /Users/zliu/src/zendesk/samson/app/models/job_execution.rb:212:in `block in commands'
/Users/zliu/src/zendesk/samson/app/models/job_execution.rb:210:in `each'
/Users/zliu/src/zendesk/samson/app/models/job_execution.rb:210:in `map'
/Users/zliu/src/zendesk/samson/app/models/job_execution.rb:210:in `commands'
/Users/zliu/src/zendesk/samson/app/models/job_execution.rb:148:in `execute!'
/Users/zliu/src/zendesk/samson/app/models/job_execution.rb:118:in `block in run!'
```

The value of `REVISION` is `nil`. It comes from `job.commit`, which happens to be `nil` if the job is created from `stage.create_deploy`.

- supply commit to job when creating a deploy for a stage
- make commit_from_git return long hash when given a hash
- only update git references of a job when either commit or tag is not nil

/cc @zendesk/samson @zendesk/vulcan

### Tasks
 - [ ] :+1: from team

### References
 - Jira link:

### Risks
- Level: Low

### CODE REVIEW

**Comment by grosser** [comment]
can we rather make sure it is always set ? / make stage.create_deploy set it ?

**Comment by zehua** [comment]
> can we rather make sure it is always set ? / make stage.create_deploy set it ?

yeah, that's what https://github.com/zendesk/samson/pull/1097/files#diff-0a8601dff8aa30a2fcff0101fc900fceL97 does. I think this was the only case with `job.commit` not set. All other code paths that create `job` are good.

File: app/models/stage.rb [hunk 1]
```diff
@@ -94,7 +94,7 @@ def confirm_before_deploying?
 
   def create_deploy(user, attributes = {})
     deploys.create(attributes.merge(release: !no_code_deployed)) do |deploy|
-      deploy.build_job(project: project, user: user, command: script)
+      deploy.build_job(project: project, user: user, command: script, commit: deploy.reference)
```
Comment (grosser) [inline app/models/stage.rb]: but why is this needed if we always have a commit ?

Comment (grosser) [inline app/models/stage.rb]: wrongline ...

File: app/models/job_execution.rb [hunk 2]
```diff
@@ -173,7 +173,7 @@ def setup!(dir)
       return false unless @repository.setup!(dir, @reference)
       commit = @repository.commit_from_ref(@reference, length: nil)
       tag = @repository.tag_from_ref(@reference)
-      @job.update_git_references!(commit: commit, tag: tag)
+      @job.update_git_references!(commit: commit, tag: tag) if commit || tag
```
Comment (grosser) [inline app/models/job_execution.rb]: but why is this needed if we always have a commit ?

**Comment by grosser** [comment]
stage/git_repo changes look good, just confused by the `if commit` change

File: app/models/git_repository.rb [hunk 3]
```diff
@@ -47,7 +47,7 @@ def clone!(from: repository_url, to: repo_cache_dir, mirror: false)
 
   def commit_from_ref(git_reference, length: 7)
     return unless ensure_local_cache!
-    command = ['git', 'describe', '--long', '--tags', '--all', "--abbrev=#{length || 40}", git_reference]
+    command = ['git', 'describe', '--always', '--long', '--tags', '--all', "--abbrev=#{length || 40}", git_reference]
```
Comment (zehua) [inline app/models/git_repository.rb]: This method seems to try to obtain the full commit hash. Why don't we use `git rev-parse`?

File: app/models/job_execution.rb (hunk 2 above)
Comment (zehua) [inline app/models/job_execution.rb]: I added this before I made the change to `commit_from_ref` in an attempt to avoid overwriting `commit` to `nil`.

It might still help in case the user passes in an invalid git ref, causing `@repository.commit_from_ref` above to return `nil`.

File: app/models/git_repository.rb (hunk 3 above)
Comment (grosser) [inline app/models/git_repository.rb]: haha yeah ... not sure why all this foobar is here, change away :D

File: app/models/job_execution.rb (hunk 2 above)
Comment (zehua) [inline app/models/job_execution.rb]: ok, invalid commit hash will not even get to this step, as it will fail at the `checkout` step https://github.com/zendesk/samson/blob/51fc8afd30746b3dfba730431da686da9c8fda0d/app/models/git_repository.rb#L24

I will undo this change.

File: app/models/git_repository.rb (hunk 3 above)
Comment (zehua) [inline app/models/git_repository.rb]: done

File: app/models/job_execution.rb (hunk 2 above)
Comment (zehua) [inline app/models/job_execution.rb]: done

Comment (grosser) [inline app/models/job_execution.rb]: pusssssh it :D

Comment (grosser) [inline app/models/job_execution.rb]: great ... github shoed me 7 hour old cache :/

File: app/models/git_repository.rb [hunk 4]
```diff
@@ -45,11 +45,10 @@ def clone!(from: repository_url, to: repo_cache_dir, mirror: false)
   end
   add_method_tracer :clone!
 
-  def commit_from_ref(git_reference, length: 7)
+  def commit_from_ref(git_reference)
```
Comment (grosser) [inline app/models/git_repository.rb]: good catch on the param never being used ❤️
